adapter/internal/grpc-client/client-pool/pool.go

```diff
@@ -0,0 +1,175 @@
+}
+
+
+func Init(serverAddress string, maxCapacity int, desiredCapacity int, dialOptions []grpc.DialOption, retryPolicy RetryPolicy) (*Pool, error) {
+	connectionPool := Pool{lock: &sync.Mutex{}};
+	connectionPool.lock.Lock();
+	defer connectionPool.lock.Unlock();
+	connectionPool.maxCapacity = maxCapacity;
+	connectionPool.desiredCapacity =  desiredCapacity;
+	connectionPool.serverAddress = serverAddress;
+	connectionPool.dialOptions = dialOptions;
+	connectionPool.availableConnections = &[]grpc.ClientConn{};
+	connectionPool.usedConnections = &[]grpc.ClientConn{};
+	connectionPool.retryPolicy = retryPolicy;
+	for i := 0; i < desiredCapacity; i++ {
+		conn, err := createGRPCConnection(connectionPool);
+		if (err != nil) {
+			return nil,err;
+		}
+		
+		availableConnections := connectionPool.availableConnections;
+		*availableConnections = append(*availableConnections, *conn);
+	}
+	return &connectionPool, nil;
+}
+
+func InitWithConfig()  (*Pool, error) {
+	conf, _ := config.ReadConfigs()
+	address := conf.Adapter.GRPCClient.ManagementServerAddress + ":" + strconv.Itoa(conf.Adapter.GRPCClient.ManagementServerGRPCPort);
+	return Init(address, conf.Adapter.GRPCClient.MaxCapacity, conf.Adapter.GRPCClient.DesiredCapacity, 
+		[]grpc.DialOption{
+			grpc.WithInsecure(),
+			grpc.WithBlock()}, 
+		RetryPolicy{
+			MaxAttempts : conf.Adapter.GRPCClient.MaxAttempts,
+			BackOffInMilliSeconds : conf.Adapter.GRPCClient.BackOffInMilliSeconds,
+			RetryableStatuses : []string{},
+		})
+}
+
+func (connectionPool *Pool) GetConnection() (*grpc.ClientConn, error){
+	connectionPool.lock.Lock();
+	defer connectionPool.lock.Unlock();
+	availableConnections := *connectionPool.availableConnections;
+	availableConnectionLength := len(availableConnections);
+	if (availableConnectionLength > 0) {
+		availableConnection := &availableConnections[0];
+		*connectionPool.availableConnections = (*connectionPool.availableConnections)[1:]
+		*connectionPool.usedConnections = append(*connectionPool.usedConnections, *availableConnection);
+		return availableConnection, nil;
+	} else {
+		totalConnectionLength := availableConnectionLength + len(*connectionPool.usedConnections);
+		if (totalConnectionLength < connectionPool.maxCapacity) {
+			return createGRPCConnection(*connectionPool);
+		} else {
+			return nil, errors.New("Maximum connection reached in the pool.")
+		}
+	}
+}
+
+func createGRPCConnection(connectionPool Pool) (*grpc.ClientConn, error) {
+	return grpc.Dial(
+		connectionPool.serverAddress, 
+		connectionPool.dialOptions...	
+	)
+}
+
+func (connectionPool *Pool) Close(connection *grpc.ClientConn) error{
+	connectionPool.lock.Lock();
+	defer connectionPool.lock.Unlock();
+	var index int = -1;
+	var connections *[]grpc.ClientConn;
+	for k, v := range *connectionPool.usedConnections {
+       if connection == &v {
+           index = k;
+		   connections = connectionPool.usedConnections;
+		   break;
+       }
+    }
+	if (index == -1) {
```

Comment:
Please improve the code readability by adding comments and giving meaningful var names. Confused about what you were trying to do.